buffer/src/main/java/io/netty/buffer/AbstractReferenceCountedByteBuf.java

```diff
@@ -133,6 +131,7 @@ private boolean release0(int decrement) {
         int rawCnt = nonVolatileRawCnt(), realCnt = toLiveRealCnt(rawCnt, decrement);
         if (decrement == realCnt) {
             if (refCntUpdater.compareAndSet(this, rawCnt, 1)) {
+                maxCapacity(-rawMaxCapacity());
```

Comment:
should we assert that this value will never be positive ?

Reply:
@normanmaurer I pushed a new commit which always updates with negative abs value here, hopefully that's just as good? Also I realized the `maxCapacity == 0` case wasn't handled, added that too. Still all tentative though pending the other investigation :)